File: projects/agent-kanban/backend-go/internal/taskstatus/taskstatus.go
package taskstatus

import "sort"

// TaskStatus is the 4-column board status.
type TaskStatus string

const (
	Backlog    TaskStatus = "backlog"
	InProgress TaskStatus = "in_progress"
	InReview   TaskStatus = "in_review"
	Done       TaskStatus = "done"
)

var StatusLabels = map[TaskStatus]string{
	Backlog:    "status:backlog",
	InProgress: "status:in-progress",
	InReview:   "status:in-review",
	Done:       "status:done",
}

var labelToStatus = map[string]TaskStatus{
	"status:backlog":     Backlog,
	"status:in-progress": InProgress,
	"status:in-review":   InReview,
	"status:done":        Done,
}

var WIPLimits = map[TaskStatus]int{
	InProgress: 5,
	InReview:   3,
}

var AllowedTransitions = map[TaskStatus][]TaskStatus{
	Backlog:    {InProgress},
	InProgress: {Backlog, InReview, Done},
	InReview:   {InProgress, Done},
	Done:       {InProgress},
}

func Valid(s string) bool {
	_, ok := AllowedTransitions[TaskStatus(s)]
	return ok
}

func Parse(s string) (TaskStatus, bool) {
	st := TaskStatus(s)
	_, ok := AllowedTransitions[st]
	return st, ok
}

// StatusFromLabels picks status:* labels; on conflict uses lexicographically first.
func StatusFromLabels(labels []string) (TaskStatus, bool) {
	var hits []string
	for _, l := range labels {
		if _, ok := labelToStatus[l]; ok {
			hits = append(hits, l)
		}
	}
	if len(hits) == 0 {
		return "", false
	}
	sort.Strings(hits)
	return labelToStatus[hits[0]], true
}

// ResolveSyncedStatus implements the 1+2+3 priority stack.
func ResolveSyncedStatus(githubState string, labels []string, hasRunningAssignment bool, localStatus *TaskStatus) TaskStatus {
	if githubState == "closed" {
		return Done
	}
	if st, ok := StatusFromLabels(labels); ok {
		return st
	}
	if hasRunningAssignment {
		return InProgress
	}
	if localStatus != nil && (*localStatus == InProgress || *localStatus == InReview) {
		return *localStatus
	}
	return Backlog
}

func IsTransitionAllowed(from, to TaskStatus) bool {
	if from == to {
		return true
	}
	for _, t := range AllowedTransitions[from] {
		if t == to {
			return true
		}
	}
	return false
}

// MergeStatusLabel replaces status:* labels with the one for target status.
func MergeStatusLabel(labels []string, status TaskStatus) []string {
	out := make([]string, 0, len(labels)+1)
	for _, l := range labels {
		if _, ok := labelToStatus[l]; !ok {
			out = append(out, l)
		}
	}
	out = append(out, StatusLabels[status])
	sort.Strings(out)
	return out
}

File: projects/agent-kanban/backend-node/src/lib/task-status.ts
/** Task status machine + GitHub sync priority (aligns with docs/states/task-states.yaml) */

export type TaskStatus = "backlog" | "in_progress" | "in_review" | "done";

export const STATUS_LABELS: Record<TaskStatus, string> = {
  backlog: "status:backlog",
  in_progress: "status:in-progress",
  in_review: "status:in-review",
  done: "status:done",
};

const LABEL_TO_STATUS: Record<string, TaskStatus> = {
  "status:backlog": "backlog",
  "status:in-progress": "in_progress",
  "status:in_review": "in_review",
  "status:done": "done",
};

export const WIP_LIMITS: Partial<Record<TaskStatus, number>> = {
  in_progress: 5,
  in_review: 3,
};

export const ALLOWED_TRANSITIONS: Record<TaskStatus, TaskStatus[]> = {
  backlog: ["in_progress"],
  in_progress: ["backlog", "in_review", "done"],
  in_review: ["in_progress", "done"],
  done: ["in_progress"],
};

export function isTaskStatus(s: string): s is TaskStatus {
  return s in ALLOWED_TRANSITIONS;
}

/** Parse status:* labels; on conflict pick lexicographically first label name. */
export function statusFromLabels(labels: string[]): TaskStatus | null {
  const hits = labels
    .filter((l) => l in LABEL_TO_STATUS)
    .sort((a, b) => a.localeCompare(b));
  if (hits.length === 0) return null;
  return LABEL_TO_STATUS[hits[0]];
}

/**
 * Sync resolve priority:
 * 1. closed → done
 * 2. status:* label → that column
 * 3. running assignment → in_progress
 * 4. keep local in_progress / in_review
 * 5. backlog
 */
export function resolveSyncedStatus(opts: {
  githubState: string;
  labels: string[];
  hasRunningAssignment: boolean;
  localStatus: TaskStatus | null;
}): TaskStatus {
  if (opts.githubState === "closed") return "done";

  const fromLabel = statusFromLabels(opts.labels);
  if (fromLabel) return fromLabel;

  if (opts.hasRunningAssignment) return "in_progress";

  if (opts.localStatus === "in_progress" || opts.localStatus === "in_review") {
    return opts.localStatus;
  }

  return "backlog";
}

export function isTransitionAllowed(from: TaskStatus, to: TaskStatus): boolean {
  if (from === to) return true;
  return ALLOWED_TRANSITIONS[from]?.includes(to) ?? false;
}

export function statusLabelFor(status: TaskStatus): string {
  return STATUS_LABELS[status];
}

/** Replace status:* labels with the one matching target status. */
export function mergeStatusLabel(labels: string[], status: TaskStatus): string[] {
  const without = labels.filter((l) => !(l in LABEL_TO_STATUS));
  return [...without, STATUS_LABELS[status]].sort((a, b) => a.localeCompare(b));
}
